### DISCUSSION TITLE
Reader indicators

### DISCUSSION THREAD

**Discussion by bytedreamer** [bytedreamer/Aporta#18] [Ideas]
Open discussion on best way to indicate various events and states using reader indicators. Typical indicators include led lights, buzzer and text screens.

**Comment by bytedreamer** [comment]
Aporta currently uses the following reader notifications:

- **Hearbeat** Permanent Red LED that blinks black every 5 seconds
- **Access Denied** Flashes Red LED with two 2 beeps.
- **Access Granted** Changes to Green LED while door strike is activated and beeps 3 times.

Interested in feedback on the best standard notifications for access readers.

**Reply by mistial-dev** [reply]
HID readers sometimes have LCD screens.  Some readers (Veridt, IIRC) don’t, but look for messages being sent to the screen to change the LEDs.  “Insert” in the message, for example, turns off the LED on the tap, but turns the contact slot green.

**Reply by mistial-dev** [reply]
One option might be to have a class that “handles” a particular set of pdid readers.  The reader could be identified, loop through the list of handlers, and select the one with the highest priority.  That would allow devices with specific hardware to be used, as well as fall back to generic readers.

It would also potentially allow a pdcap response to be used, so you could do logic like treating readers with contact differently, or HID models with LCDs vs no LCD, or keypad vs no keypad.

**Reply by bytedreamer** [reply]
@mistial-dev Good idea having PD Capabilities drive the type of notifications. I would also add device identification needs to be included for specific reader model behavior.

**Reply by rsgmodelworks** [reply]
I think what you propose using looks fine.  I have seen people prefer green blinking slowly to indicate unlocked.  I have also seen requests for this to be configurable (by mortals, using non-buried, natural-language documentation, I'm looking at you Lenel.)

**Comment by bytedreamer** [comment]
Are there any interest in creating a scripting languages for customizing reader indicators?

**Reply by mistial-dev** [reply]
https://github.com/oleg-shilo/cs-script Is an option.

**Reply by bytedreamer** [reply]
The more I think about to handle reader notifications and all the possible permutations between reader types, scripting might be the best solution. 

I'm leaning toward PowerShell, because there is ability to restrict the commands that can be called. We would want to be careful not to allow malicious operations to be inserted into the application. https://learn.microsoft.com/en-us/powershell/scripting/developer/hosting/windows-powershell-host-quickstart?view=powershell-7.3

I believe you can pass a C# object that would include the device information and methods for sending notifications to the embedded PowerShell script. https://keithbabinec.com/2020/02/15/how-to-run-powershell-core-scripts-from-net-core-applications/

**Comment by mistial-dev** [comment]
Does that work well on Linux and Mac?

**Reply by bytedreamer** [reply]
It should, Microsoft has released a cross platform version of PowerShell. https://github.com/PowerShell/PowerShell

**Comment by mistial-dev** [comment]
Interesting.  I wouldn't normally consider that a scripting language, but it might work.
